Adding support for Terraform Destroy. Adding ConfirmApply customization

jleopold28 opened the pull request:
Support for TF destroy (https://github.com/manheim/terraform-pipeline/issues/88)

Also adding support to customize the ConfirmApply messages. (https://github.com/manheim/terraform-pipeline/issues/24)

kmanning commented on src/ConditionalApplyPlugin.groovy:
```diff
@@ -1,18 +1,22 @@
+        if (stage instanceof TerraformDestroyStage){
+            stage.decorateAround(DESTROY, onlyOnExpectedBranch())
+        }
     }
```
There's two ways you can avoid the `instanceof TerraformDestroyStage`.
1. If you define a `public void apply(TerraformDestroyStage stage)` method, Groovy will be smart enough to use that one for TerraformDestroyStages, and use this one for normal TerraformEnvironmentStages.  That is, you let the compiler implicitly do the `if instanceof TerraformDestroy` on your behalf.
2. We don't care whether or not you're a normal TerraformEnvironmentStage or a TerraformDestroyStage.  Either way, we do the same thing, and decorate the CONFIRM and APPLY.  This means that in a TerraformDestroyStage 'APPLY' is now the time that we run the `terraform destroy` command instead of the `terraform apply` command.

kmanning commented on src/ConfirmApplyPlugin.groovy:
```diff
@@ -13,19 +14,24 @@ class ConfirmApplyPlugin implements TerraformEnvironmentStagePlugin {
         if (enabled) {
-            stage.decorate(CONFIRM, addConfirmation())
+            stage.decorate(CONFIRM, addConfirmation(command_name))
         }
```
Here again, we can avoid the `instanceof`, and in fact, it'll give us an added bonus.  We could leave `apply(TerraformEnvironmentStage stage)` unmodified, and define a new `apply(TerraformDestroyStage)`.  It could get a *FULLY* custom message "*DESTORY* *DESTROY* YOU ARE GOING TO DESTROY YOUR ENVIRONMENT *DESTROY* *DESTROY*" <-- I'm on the fence as to whether that's necessary, but I am calling it out as a possibility.

kmanning commented on src/TargetPlugin.groovy:
```diff
@@ -27,6 +28,15 @@ class TargetPlugin implements TerraformPlanCommandPlugin, TerraformApplyCommandP
+               .each { item -> command.withArgument("-target ${item}") }
+    }
+
     @Override
```
Does the TargetPlugin apply in the case of destroy? If you wanted to remove a single resource without touching anything else, couldn't you use a normal pipeline, remove your resource from your code, and target just the removed resource?

If we can avoid changing plugins at all, I would love that.

kmanning commented on src/TerraformDestroyCommand.groovy:
```diff
@@ -0,0 +1,88 @@
+class TerraformDestroyCommand {
+    private boolean input = false
```
I can't say yet whether or not it might make sense to say `class TerraformDestoryCommand extends TerraformApplyCommand`?

kmanning commented on src/TerraformDestroyStage.groovy:
```diff
@@ -0,0 +1,52 @@
+                    decorations.apply("Around-${DESTROY}") {
+                        stage("${DESTROY}-${environment}") {
+                            decorations.apply(DESTROY) {
+                                sh destroyCommand.toString()
```
Here's where we could be "clever".  Instead of creating a new hook-point of "DESTROY", we *could* adapt/reuse "APPLY".  Then "everything just works" in terms of plugins, and perhaps no plugins would have to be modified to work with TerraformDestroy?

kmanning commented on src/TerraformEnvironmentStage.groovy:
```diff
@@ -1,14 +1,14 @@
-    private static globalPlugins = DEFAULT_PLUGINS.clone()
+    protected static final DEFAULT_PLUGINS = [ new ConditionalApplyPlugin(), new ConfirmApplyPlugin(), new DefaultEnvironmentPlugin() ]
+    protected static globalPlugins = DEFAULT_PLUGINS.clone()
 
```
Now that you've done that, you just highlighted the fact that many of these variables never needed to be instance variables to begin with.

Rather than make initCommand/planCommand/applyCommand protected, you could simply delete them, and make them local only to the pipelineConfiguration method.  In which case TerraformDestroy would have no idea about them (which is good) and would simply define it's own.

environment already has a public getEnvironment() method, it would not need to be made public.  jenkinsfile can be accessed globally, it too would not need to be made protected.

applyPlugins manipulates the plugins as needed, none of the plugin variables would need to be made protected.

kmanning replied:
I feel like you've highlighted a code smell as well.  I think you're showing that maybe pipelineConfiguration should perhaps be implemented in some other way.  In fact, you might have just hit the nail on the head.  If pipelineConfiguration were defined by a class/object that could be swapped out, that might in fact solve *everything*.  We wouldn't need a new TerraformDestoryStage at all.

kmanning commented on src/ConfirmApplyPlugin.groovy:
```diff
@@ -14,18 +15,18 @@ class ConfirmApplyPlugin implements TerraformEnvironmentStagePlugin {
-                        ok: 'Run terraform APPLY now',
+                        message: "Are you absolutely sure the plan above is correct, and should be IMMEDIATELY DEPLOYED via \"terraform ${command_name}\"?",
+                        ok: "Run terraform ${command_name.toUpperCase()} now",
                         submitterParameter: 'approver'
```
Ooohh, this is super clever.  What if  we took it a  step further?

```
class ConfirmApplyPlugin...
   ...
   private  confirmMessage = 'Are you absolutely sure...' <--  the stuff that's currently hard coded.
   ...
        input (
            message: confirmMessage,
            ok: okMessage,
            submitterParameter: submitterParameter
        )
    ....
    public static withConfirmMessage(String newMessage) { this.confirmMessage = newMessage }
    public static withOkMessage(String newMessage) { this.okayMessage = newMessage }
    public static withSubmitterParameter(String newParam) { this.submitterParameter = newParam }
```

Now this has nothing to do with the DestroyPlugin.  *ANY* plugin that may need to manipulate confirmations can do so.

kmanning commented on src/DestroyPlugin.groovy:
```diff
@@ -0,0 +1,16 @@
+        stage.withStrategy(new DestroyStrategy())
+    }
+
+}
```
What you did above gave me a new idea!! What if...

```
class DestroyPlugin implements TerraformEnvironmentStagePlugin {

    public static void init() {
        DestroyPlugin plugin = new DestroyPlugin()

        // Need to set command name to "destroy"
        ConfirmApplyPlugin.withConfirmationMessage("YOU ARE ABOUT TO DESTORY YOUR ENVIORNMENT")
        TerraformPlanCommand.withDestroy()
        TerraformApplyCommand.withDestroy()
    }
}
```

```
class TerraformPlanCommand {
...
    public static withDestroy() {
        withArgument('-destory')
    }
}
```

```
class TerraformApplyCommand {
...
    public static withDestroy() {
         this.command = 'destroy'
    }
}
```

1. We would no longer need to modify TerraformEnvironmentStage *at all*.  
2. We wouldn't need TerraformDestroyStage.  
3. We wouldn't need a new TerraformDestoryCommand
4. *ALL* existing plugins would "just work"

kmanning replied:
The above could even be implemented without adding new `withDestroys()` methods.  And in fact, not creating these new methods might be the better/more-conservative approach.

vincentclee commented:
This is some great work @jleopold28 ! Thoughts on adding `-refresh=false` to the default options for destroy?

**Common Scenario**
https://github.com/hashicorp/terraform/issues/15386#issuecomment-588295351

@kmanning

jleopold28 commented:
@vincentclee thanks for testing! I don't think `-refresh=false` should be a default option, but I am in favor of allowing it to be configured.

kmanning commented on docs/DestroyPlugin.md:
````diff
@@ -0,0 +1,73 @@
+        .then(destroyProd)
+        .build()
+```
+
````
I'm not sure I'm a fan of this.  The premise of the library is that once you  call `.build()`, no other code should be executed.  I think it might be a necessary constraint in order to be to effectively create Declarative Pipelines (the syntax is really strict).  Note: [README](https://github.com/manheim/terraform-pipeline#how-to-use) `The design of this library is influenced by the Builder Pattern - your pipeline has been configured, but hasn't been constructed just yet. Finalize and create your pipeline by calling the build() method. This should only be done once - no code should come after calling this method.`


At any rate, I don't think I'm ready to concede on the `.build()` rule just yet.  

Can we  separate this PR into UseCases?  There's 1 UseCase of just terminating all the environments, which is your first example.  Can we do a separate PR to implement the other UseCase of deploying+destroying?  I'd like to talk more about how we  could implement that.

kmanning commented on src/ConditionalApplyPlugin.groovy:
```diff
@@ -13,6 +14,7 @@ public class ConditionalApplyPlugin implements TerraformEnvironmentStagePlugin {
         stage.decorateAround(CONFIRM, onlyOnExpectedBranch())
         stage.decorateAround(APPLY, onlyOnExpectedBranch())
+        stage.decorateAround(DESTROY, onlyOnExpectedBranch())
     }
```
If we only attack UseCase 1 in this PR (destroy-only), we no longer need these changes.

kmanning commented on src/DefaultStrategy.groovy:
```diff
@@ -0,0 +1,53 @@
+            }
+        }
+    }
+}
```
Same as above: "If we only attack UseCase 1 in this PR (destroy-only), we no longer need these changes."

kmanning commented on src/DestroyStrategy.groovy:
```diff
@@ -0,0 +1,53 @@
+            }
+        }
+    }
+}
```
Same "If we only attack UseCase 1 in this PR (destroy-only), we no longer need these changes."

kmanning commented on src/TerraformEnvironmentStage.groovy:
```diff
@@ -46,46 +45,13 @@ class TerraformEnvironmentStage implements Stage {
-            }
-        }
+        return strategy.createPipelineClosure(environment, decorations)
     }
```
Same "If we only attack UseCase 1 in this PR (destroy-only), we no longer need these changes."

kmanning commented:
@vincentclee @jleopold28 ?

In regards to the `-refresh=false`, does anything need to be done from a code  perspective?  Would `TerraformApply.withArgument('-refresh=false')` work?

kmanning commented:
Oh!  I also finally published the new Contributor instructions.  Let's do away with long-lived dev branches, and queue work in master instead.  It's simpler that way.  (Ie: you can make your PR against master, and I'll delete v5.9-dev - there's nothing in there at the moment).  https://github.com/manheim/terraform-pipeline#how-to-contribute

jleopold28 commented:
@kmanning I updated the base to `master`. I also removed the section in the docs talking about "Destroy after Deploy". We can discuss that as a future piece of work.

jleopold28 commented:
> @vincentclee @jleopold28 ?
> 
> In regards to the `-refresh=false`, does anything need to be done from a code perspective? Would `TerraformApply.withArgument('-refresh=false')` work?

That would work. However, I don't think that should be the default behavior...

kmanning commented:
You don't think "-refresh=false" should be the default behavior, or you don't think users should *need* to use `.withArgument` in order to get this to work?

jleopold28 commented:
> You don't think "-refresh=false" should be the default behavior, or you don't think users should _need_ to use `.withArgument` in order to get this to work?

@kmanning I think users should have to use `withArgument` or something similar to add that option. 

When doing a destroy, I think we should always refresh the state before executing a destroy. If users are hitting a bug that requires them to add `refresh=false`, they should specify that with arguments during the plugin init.

kmanning commented on src/TerraformEnvironmentStage.groovy:
```diff
@@ -14,6 +12,7 @@ class TerraformEnvironmentStage implements Stage {
     public static final String PLAN = 'plan'
     public static final String CONFIRM = 'confirm'
     public static final String APPLY = 'apply'
+    public static final String DESTROY = 'destroy'
```
This new state of "DESTROY".  How will this behavior with the existing plugins that are part of  the library (or worse, plugins that are in completely external libraries?)

eg: 

* https://github.com/manheim/terraform-pipeline/blob/74de97b89fdedf087fde028686b1e6530a30aed2/src/ParameterStoreExecPlugin.groovy#L14-L20
* https://github.com/manheim/terraform-pipeline/blob/74de97b89fdedf087fde028686b1e6530a30aed2/src/AnsiColorPlugin.groovy#L11-L14
* https://github.com/manheim/terraform-pipeline/blob/817be51f85223d535120fb7d17e35698f34d8f17/src/ParameterStoreBuildWrapperPlugin.groovy#L17-L29
* https://github.com/manheim/terraform-pipeline/blob/00e3eed6cfcf8e4ad40a68a323f116202cda90da/src/CrqPlugin.groovy#L11-L16

jleopold28 replied:
@kmanning I am making some changes now. I will have to add `stage.decorate(DESTROY, someMethod(options))` to the above plugins. This will add functionality if DESTROY is enabled. Otherwise this will not change the existing functionality.

kmanning commented on docs/DestroyPlugin.md:
```diff
@@ -0,0 +1,50 @@
+When using this plugin, your pipeline will look something like this:
+
+![DestroyPluginPipeline](../images/destroy-pipeline.png)
+
```
Nice touch on the screenshot - thank you for that.

kmanning commented:
Thanks for accounting for the other plugins.  I'm still concerned about the maintainability of the existing PLAN/APPLY hooks, and I'm concerned about introducing more problems with the new DESTROY hook.  But I think that can be tackled in a separate PR/refactor.

kmanning commented:
:shipit:
